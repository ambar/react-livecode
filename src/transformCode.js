import recast from 'recast'

const transform = ast => {
  const b = recast.types.builders
  let expressionStatements = 0
  const newAst = {
    ...ast,
    body: ast.body.map(node => {
      if (node.type === 'ExpressionStatement') {
        expressionStatements += 1
        return b.expressionStatement(
          b.callExpression(b.identifier('__onexpression__'), [node.expression])
        )
      }
      return node
    }),
  }
  return [expressionStatements, newAst]
}

const transformCode = code => {
  const [expressionStatements, newAst] = transform(recast.parse(code).program)
  return [expressionStatements, recast.print(newAst).code]
}

export default transformCode
