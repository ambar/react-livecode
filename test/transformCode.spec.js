import transformCode from '../src/transformCode'

describe('transformCode', () => {
  it('should add `onexpression`', () => {
    expect(
      transformCode(`
        2 ** 3
        chunk(['a', 'b', 'c', 'd'], 2)
        new Map([['key1', 'value1'], ['key2', 'value2']])
        const never = new Promise(_ => {})
        never
    `)
    ).toMatchSnapshot()
  })
})
