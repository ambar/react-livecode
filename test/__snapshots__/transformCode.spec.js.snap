// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformCode should add \`onexpression\` 1`] = `
Array [
  4,
  "__onexpression__(2 ** 3);
__onexpression__(chunk(['a', 'b', 'c', 'd'], 2));
__onexpression__(new Map([['key1', 'value1'], ['key2', 'value2']]));
const never = new Promise(_ => {})
__onexpression__(never);",
]
`;
